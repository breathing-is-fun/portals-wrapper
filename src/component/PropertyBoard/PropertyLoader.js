/*
 * @Author: zy9@github.com/zy410419243
 * @Date: 2018-10-14 10:31:52
 * @Last Modified by: zy9
 * @Last Modified time: 2018-10-29 14:08:56
 */

// 获得style中文枚举
export default class PropertyLoader {
	load = (enumDatas, style = {}, key, callback) => {
		// fetch('../../../mock/propertyDatas.json')
		// 	.then(result => result.json())
		// 	.then(result => {
		// 		const propertyDatas = this.hanldePropertyEnum(style, result.data);

		// 		callback && callback(propertyDatas);
		// 	});
		const propertyDatas = this.hanldePropertyEnum(style, enumDatas, key);

		callback && callback(propertyDatas);
	}

	hanldePropertyEnum = (style, enumDatas, id) => {
		let result = [];
		const styleKeys = Object.keys(style);

		for(let styleKey of styleKeys) {
			for(let item of enumDatas) {
				const { text, key, isStyle = true } = item;

				if(key == styleKey) {
					result.push({ text, key, isStyle, value: style[styleKey], id });
				}
			}
		}

		return result;
	}
}