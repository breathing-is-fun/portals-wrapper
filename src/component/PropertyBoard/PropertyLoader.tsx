// 获得style中文枚举
export default class PropertyLoader {
  load = (
    enumDatas: Array<any>,
    style = {},
    key: string,
    callback: (propertyDatas: any) => void,
  ) => {
    const propertyDatas = this.hanldePropertyEnum(style, enumDatas, key);
    callback && callback(propertyDatas);
  };

  hanldePropertyEnum = (
    style: any,
    enumDatas: Array<any>,
    id: string | number,
  ) => {
    let result = [];
    const styleKeys = Object.keys(style);

    for (let styleKey of styleKeys) {
      for (let item of enumDatas) {
        const { text, key, isStyle = true } = item;

        if (key == styleKey) {
          result.push({
            text,
            key,
            isStyle,
            value: style[styleKey],
            id,
          });
        }
      }
    }

    return result;
  };
}
