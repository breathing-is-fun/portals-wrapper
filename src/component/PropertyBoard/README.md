English | [简体中文](./README-zh_CN.md)

# 简介

- 属性面板，用作修改套餐编辑时的外壳属性

- 基于 [antd的Drawer](https://ant.design/components/drawer-cn/)
  
# API  

| 名称 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| visible | Drawer 是否可见 | boolean | - |
| onClose | 点击遮罩层或右上角叉或取消按钮的回调 | function(e) | - |
| shellStyleDatas | 把样式属性根据 `enumDatas` 变成中文，`i` 用作标识当前操作的 `Shell` | [{ i: string, style: {} }] | - |
| enumDatas | 枚举，比如 `backgroundColor` => `背景颜色`，[详情在这里](#enumDatas的对象) | [{ text: string, key: string, isStyle: boolean }] | [] |

## enumDatas的对象

| 名称 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| key | 属性名，需要写成 [`inline style`](https://reactjs.org/docs/dom-elements.html#style) | string | - |
| text | 属性中文名 | string | - |
| isStyle | 当需要修改不是css的参数比如 `title` 时，可以传 `false` 跳出遍历 | boolean | - |

# 示例

待填坑