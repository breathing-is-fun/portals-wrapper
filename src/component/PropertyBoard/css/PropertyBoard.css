.PropertyBoard ul {
    list-style: none;
}

.PropertyBoard .property-wrapper .property-key {
    text-align: right;
    vertical-align: middle;
    line-height: 40px;
    display: inline-block;
    overflow: hidden;
    white-space: nowrap;
    width: 25%;
}

.PropertyBoard .property-wrapper .property-value {
    display: inline-block;
    -webkit-box-sizing: border-box;
    box-sizing: border-box;
    width: 75%;
}