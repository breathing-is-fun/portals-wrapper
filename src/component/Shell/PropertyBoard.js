/*
 * @Author: zy9@github.com/zy410419243
 * @Date: 2018-10-10 10:12:51
 * @Last Modified by: zy9
 * @Last Modified time: 2018-10-10 16:15:23
 */
import React, { Component } from 'react';

import { Drawer, Icon, Form, Input } from 'antd';
const FormItem = Form.Item;

class PropertyBoard extends Component {
	constructor (props) {
		super(props);

		this.state = {

		};
	}

    componentDidMount = () => {

    }

	handleSubmit = e => {
		e.preventDefault();

		this.props.form.validateFieldsAndScroll((err, values) => {
			if (!err) {
			  	console.log('Received values of form: ', values);
			}
		});
	}

    render = () => {
    	const { visible, onClose, dataSource, form } = this.props;
    	const { style, title } = dataSource;
    	const { getFieldDecorator  } = form;

    	const drawerTitle = (
    		<div>
    			<Icon type='form' theme='outlined' />
    			<span style={{ paddingLeft: 10 }}>{ `${ title } 外壳属性修改` }</span>
    		</div>
    	);

    	const drawerProps = {
    		visible,
    		className: 'PropertyBoard',
    		onClose: () => onClose && onClose(!visible),
    		title: drawerTitle,
    		width: '20%',
    	};

    	const formItemLayoutProps = {
    		labelCol: {
    			xs: { span: 24 },
    			sm: { span: 4 },
    		},
    		wrapperCol: {
    			xs: { span: 24 },
    			sm: { span: 18 },
    		},
    	};

    	return (
    		<div className='PropertyBoard'>
    			<Drawer { ...drawerProps }>
    				<Form onSubmit={ this.handleSubmit }>
    					<FormItem { ...formItemLayoutProps } label='标题'>
    						{
    							getFieldDecorator('email', {
    								rules: [{
    									type: 'email', message: 'The input is not valid E-mail!',
    								}, {
    									required: true, message: 'Please input your E-mail!',
    								}],
    							})(<Input />)
    						}
    					</FormItem>
    				</Form>
    			</Drawer>
    		</div>
    	);
    }
}

export default Form.create()(PropertyBoard);