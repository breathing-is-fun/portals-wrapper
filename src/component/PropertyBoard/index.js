/*
 * @Author: zy9@github.com/zy410419243
 * @Date: 2018-10-10 10:12:51
 * @Last Modified by: zy9
 * @Last Modified time: 2018-10-29 14:10:04
 */
import React, { Component } from 'react';

import { Drawer, Icon, Input } from 'antd';
import Loader from './PropertyLoader';

import './css/PropertyBoard.css';

export default class PropertyBoard extends Component {
	constructor (props) {
		super(props);

		this.state = {
			propertyDatas: [],
		};

		this.loader = new Loader();
		this.currentShellStyle = {};
	}

    componentDidMount = () => {
    	this.loadPropertyDatas(this.props.visible);
    }

	loadPropertyDatas = visible => {
		if(visible) {
			const { shellStyleDatas, enumDatas } = this.props;
			const { i: key, style } = shellStyleDatas;

			this.loader.load(enumDatas, style, key, propertyDatas => this.setState({ propertyDatas }));
		}
	}

	handleInput = (value, key, id) => {
		const { onChange } = this.props;

		this.currentShellStyle[key] = value;
		this.currentShellStyle.id = id;

		onChange && onChange(this.currentShellStyle);
	}

    render = () => {
    	const { visible, onClose, shellStyleDatas } = this.props;
    	const { title } = shellStyleDatas;
    	const { propertyDatas } = this.state;

    	const drawerTitle = (
    		<div>
    			<Icon type='form' theme='outlined' />
    			<span style={{ paddingLeft: 10 }}>{ `${ title } 外壳属性修改` }</span>
    		</div>
    	);

    	const drawerProps = {
    		visible,
    		className: 'PropertyBoard',
    		onClose: () => onClose && onClose(!visible),
    		title: drawerTitle,
    		width: '20%',
    		destroyOnClose: true,
    	};

    	return (
    		<div className='PropertyBoard'>
    			<Drawer { ...drawerProps }>
    				<ul className='property-wrapper'>
    					{
    						propertyDatas.map(item => {
    							const { key, text, value, id } = item;

    							return (
    								<li key={ `property-li-${ key }` }>
    									<div className='property-key'>{ text }：</div>
    									<div className='property-value'>
    										<Input onBlur={ e => this.handleInput(e.target.value, key, id) } defaultValue={ value } />
    									</div>
    								</li>
    							);
    						})
    					}
    				</ul>
    			</Drawer>
    		</div>
    	);
    }
}