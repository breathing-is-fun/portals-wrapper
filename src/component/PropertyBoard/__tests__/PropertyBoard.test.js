import React from 'react';
import { render, mount, shallow } from 'enzyme';
import PropertyBoard from '..';

describe('PropertyBoard', () => {
  it('render correctly', () => {
    const demo = (
      <PropertyBoard visible={true} shellStyleDatas={{}} enumDatas={[]} />
    );

    expect(mount(demo)).toMatchSnapshot();
  });
});
