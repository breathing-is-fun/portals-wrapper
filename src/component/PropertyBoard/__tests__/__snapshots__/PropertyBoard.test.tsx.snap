// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyBoard render correctly 1`] = `
<PropertyBoard
  enumDatas={Array []}
  shellStyleDatas={Object {}}
  visible={true}
>
  <div
    className="PropertyBoard"
  >
    <withConfigConsumer(Drawer)
      className="PropertyBoard"
      destroyOnClose={true}
      onClose={[Function]}
      title={
        <div>
          <Icon
            theme="outlined"
            type="form"
          />
          <span
            style={
              Object {
                "paddingLeft": 10,
              }
            }
          >
            undefined 外壳属性修改
          </span>
        </div>
      }
      visible={true}
      width="20%"
    >
      <Drawer
        className="PropertyBoard"
        closable={true}
        destroyOnClose={true}
        getPrefixCls={[Function]}
        height={256}
        level={null}
        mask={true}
        maskClosable={true}
        onClose={[Function]}
        placement="right"
        prefixCls="ant-drawer"
        renderEmpty={[Function]}
        title={
          <div>
            <Icon
              theme="outlined"
              type="form"
            />
            <span
              style={
                Object {
                  "paddingLeft": 10,
                }
              }
            >
              undefined 外壳属性修改
            </span>
          </div>
        }
        visible={true}
        width="20%"
      >
        <Drawer
          className="PropertyBoard"
          closable={true}
          destroyOnClose={true}
          duration=".3s"
          ease="cubic-bezier(0.78, 0.14, 0.15, 0.86)"
          firstEnter={false}
          getContainer="body"
          getPrefixCls={[Function]}
          handler={false}
          level={null}
          mask={true}
          maskClosable={true}
          maskStyle={Object {}}
          onChange={[Function]}
          onClose={[Function]}
          onHandleClick={[Function]}
          onMaskClick={[Function]}
          open={true}
          placement="right"
          prefixCls="ant-drawer"
          renderEmpty={[Function]}
          showMask={true}
          style={
            Object {
              "transform": undefined,
              "zIndex": undefined,
            }
          }
          title={
            <div>
              <Icon
                theme="outlined"
                type="form"
              />
              <span
                style={
                  Object {
                    "paddingLeft": 10,
                  }
                }
              >
                undefined 外壳属性修改
              </span>
            </div>
          }
          visible={true}
          width="20%"
          wrapperClassName=""
        >
          <Portal
            containerInfo={
              <div>
                <div
                  class="ant-drawer ant-drawer-right ant-drawer-open PropertyBoard"
                >
                  <div
                    class="ant-drawer-mask"
                  />
                  <div
                    class="ant-drawer-content-wrapper"
                    style="width: 20%;"
                  >
                    <div
                      class="ant-drawer-content"
                    >
                      <div
                        class="ant-drawer-wrapper-body"
                        style="overflow: auto; height: 100%;"
                      >
                        <div
                          class="ant-drawer-header"
                        >
                          <div
                            class="ant-drawer-title"
                          >
                            <div>
                              <i
                                aria-label="icon: form"
                                class="anticon anticon-form"
                              >
                                <svg
                                  aria-hidden="true"
                                  class=""
                                  data-icon="form"
                                  fill="currentColor"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M904 512h-56c-4.4 0-8 3.6-8 8v320H184V184h320c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V520c0-4.4-3.6-8-8-8z"
                                  />
                                  <path
                                    d="M355.9 534.9L354 653.8c-.1 8.9 7.1 16.2 16 16.2h.4l118-2.9c2-.1 4-.9 5.4-2.3l415.9-415c3.1-3.1 3.1-8.2 0-11.3L785.4 114.3c-1.6-1.6-3.6-2.3-5.7-2.3s-4.1.8-5.7 2.3l-415.8 415a8.3 8.3 0 0 0-2.3 5.6zm63.5 23.6L779.7 199l45.2 45.1-360.5 359.7-45.7 1.1.7-46.4z"
                                  />
                                </svg>
                              </i>
                              <span
                                style="padding-left: 10px;"
                              >
                                undefined 外壳属性修改
                              </span>
                            </div>
                          </div>
                          <button
                            aria-label="Close"
                            class="ant-drawer-close"
                          >
                            <i
                              aria-label="icon: close"
                              class="anticon anticon-close"
                            >
                              <svg
                                aria-hidden="true"
                                class=""
                                data-icon="close"
                                fill="currentColor"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                                />
                              </svg>
                            </i>
                          </button>
                        </div>
                        <div
                          class="ant-drawer-body"
                        >
                          <ul
                            class="property-wrapper"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            }
          >
            <div
              className="ant-drawer ant-drawer-right ant-drawer-open PropertyBoard"
              onTransitionEnd={[Function]}
              style={
                Object {
                  "transform": undefined,
                  "zIndex": undefined,
                }
              }
            >
              <div
                className="ant-drawer-mask"
                onClick={[Function]}
                style={Object {}}
              />
              <div
                className="ant-drawer-content-wrapper"
                style={
                  Object {
                    "height": undefined,
                    "msTransform": "",
                    "transform": "",
                    "width": "20%",
                  }
                }
              >
                <div
                  className="ant-drawer-content"
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                >
                  <div
                    className="ant-drawer-wrapper-body"
                    onTransitionEnd={[Function]}
                    style={
                      Object {
                        "height": "100%",
                        "overflow": "auto",
                      }
                    }
                  >
                    <div
                      className="ant-drawer-header"
                    >
                      <div
                        className="ant-drawer-title"
                      >
                        <div>
                          <Icon
                            theme="outlined"
                            type="form"
                          >
                            <LocaleReceiver
                              componentName="Icon"
                            >
                              <i
                                aria-label="icon: form"
                                className="anticon anticon-form"
                              >
                                <IconReact
                                  className=""
                                  type="form-o"
                                >
                                  <svg
                                    aria-hidden="true"
                                    className=""
                                    data-icon="form"
                                    fill="currentColor"
                                    height="1em"
                                    key="svg-form"
                                    viewBox="64 64 896 896"
                                    width="1em"
                                  >
                                    <path
                                      d="M904 512h-56c-4.4 0-8 3.6-8 8v320H184V184h320c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V520c0-4.4-3.6-8-8-8z"
                                      key="svg-form-svg-0"
                                    />
                                    <path
                                      d="M355.9 534.9L354 653.8c-.1 8.9 7.1 16.2 16 16.2h.4l118-2.9c2-.1 4-.9 5.4-2.3l415.9-415c3.1-3.1 3.1-8.2 0-11.3L785.4 114.3c-1.6-1.6-3.6-2.3-5.7-2.3s-4.1.8-5.7 2.3l-415.8 415a8.3 8.3 0 0 0-2.3 5.6zm63.5 23.6L779.7 199l45.2 45.1-360.5 359.7-45.7 1.1.7-46.4z"
                                      key="svg-form-svg-1"
                                    />
                                  </svg>
                                </IconReact>
                              </i>
                            </LocaleReceiver>
                          </Icon>
                          <span
                            style={
                              Object {
                                "paddingLeft": 10,
                              }
                            }
                          >
                            undefined 外壳属性修改
                          </span>
                        </div>
                      </div>
                      <button
                        aria-label="Close"
                        className="ant-drawer-close"
                        onClick={[Function]}
                      >
                        <Icon
                          type="close"
                        >
                          <LocaleReceiver
                            componentName="Icon"
                          >
                            <i
                              aria-label="icon: close"
                              className="anticon anticon-close"
                            >
                              <IconReact
                                className=""
                                type="close-o"
                              >
                                <svg
                                  aria-hidden="true"
                                  className=""
                                  data-icon="close"
                                  fill="currentColor"
                                  height="1em"
                                  key="svg-close"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                                    key="svg-close-svg-0"
                                  />
                                </svg>
                              </IconReact>
                            </i>
                          </LocaleReceiver>
                        </Icon>
                      </button>
                    </div>
                    <div
                      className="ant-drawer-body"
                    >
                      <ul
                        className="property-wrapper"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </Portal>
        </Drawer>
      </Drawer>
    </withConfigConsumer(Drawer)>
  </div>
</PropertyBoard>
`;
